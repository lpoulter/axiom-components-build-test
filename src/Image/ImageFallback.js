import PropTypes from 'prop-types';
import { Component } from 'react';

const cancelable = (fn) => {
  let cancelled;

  const wrapped = (...args) => {
    if (!cancelled) {
      return fn(...args);
    }
  };

  wrapped.cancel = () => cancelled = true;

  return wrapped;
};

export default class ImageFallback extends Component {
  static propTypes = {
    children: PropTypes.node,
    fallback: PropTypes.node,
    onError: PropTypes.func,
    onLoad: PropTypes.func,
    src: PropTypes.string,
  };

  constructor(props) {
    super(props);
    this.state = {
      useFallback: false,
    };
  }

  componentDidMount() {
    const { fallback, src, onError, onLoad } = this.props;

    if (src !== undefined && fallback !== undefined) {
      this.image = new window.Image();
      this.image.src = src;
      this.image.onload = cancelable((...args) => {
        this.setState(({ useFallback }) => useFallback && ({ useFallback: false }));
        if (onLoad) onLoad(...args);
      });

      this.image.onerror = cancelable((...args) => {
        this.setState(({ useFallback }) => !useFallback && ({ useFallback: true }));
        if (onError) onError(...args);
      });
    }
  }

  UNSAFE_componentWillUpdate(nextProps) {
    if (this.image && nextProps.src !== this.props.src) {
      this.image.src = nextProps.src;
    }
  }

  componentWillUnmount() {
    if (this.image) {
      this.image.onload.cancel();
      this.image.onerror.cancel();
    }
  }

  render() {
    const { children, fallback, src } = this.props;
    const { useFallback } = this.state;

    if ((!src || useFallback) && fallback) {
      return fallback;
    }

    return children;
  }
}
