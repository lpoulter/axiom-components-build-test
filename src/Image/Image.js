import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Base from '../Base/Base';
import ImageFallback from './ImageFallback';
import './Image.css';

export default class Image extends Component {
  static propTypes = {
    /** Border applied around the image */
    border: PropTypes.oneOf(['small', 'large']),
    /** Fallback content when the image fails to load */
    children: PropTypes.node,
    /** Background color of the image */
    color: PropTypes.string,
    /** Height of the image. When shape is circle this will be ignored and width will be applied. */
    height: PropTypes.string,
    /** @type {[type]} [description] */
    maxWidth: PropTypes.string,
    /** SKIP */
    onError: PropTypes.func,
    /** SKIP */
    onLoad: PropTypes.func,
    /** Shape of the image */
    shape: PropTypes.oneOf(['circle', 'rounded', 'square']),
    /** Source of the image */
    src: PropTypes.string,
    /** Width of the image. When shape is circle this will also be used for height. */
    width: PropTypes.string,
  };

  static defaultProps = {
    maxWidth: '100%',
    shape: 'square',
  };

  render() {
    const {
      border,
      children,
      color,
      height,
      maxWidth,
      onError,
      onLoad,
      shape,
      src,
      width,
      ...rest
    } = this.props;

    const style = {
      maxWidth,
      width: shape === 'circle' ? (width || height) : width,
      height: shape === 'circle' ? (width || height) : height,
      backgroundColor: color,
    };

    const classes = classnames('ax-image', `ax-image--${shape}`, {
      [`ax-image--border-${border}`]: border,
    });

    return (
      <ImageFallback
          fallback={ children }
          onError={ onError }
          onLoad={ onLoad }
          src={ src }>
        <Base
            space="x4"
            { ...rest }
            Component="img"
            className={ classes }
            src={ src }
            style={ style } />
      </ImageFallback>
    );
  }
}
