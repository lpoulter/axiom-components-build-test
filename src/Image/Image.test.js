import React from 'react';
import renderer from 'react-test-renderer';
import Image from './Image';

const getComponent = (props = {}) =>
  renderer.create(
    <Image src="/image.png" { ...props } />
  );

describe('Image', () => {
  it('renders with defaultProps', () => {
    const component = getComponent();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('renders with border', () => {
    ['small', 'large'].forEach((border) => {
      it(border, () => {
        const component = getComponent({ border });
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('renders with shape', () => {
    ['circle', 'rounded', 'square'].forEach((shape) => {
      it(shape, () => {
        const component = getComponent({ shape });
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('with shape circle', () => {
    it('uses width for both width and height when given', () => {
      const component = getComponent({ shape: 'circle', width: '3rem' });
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('uses height for both width and height when given', () => {
      const component = getComponent({ shape: 'circle', height: '3rem' });
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
