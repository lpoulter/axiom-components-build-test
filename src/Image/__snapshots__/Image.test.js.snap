// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders with border large 1`] = `
<img
  className="ax-image ax-image--square ax-image--border-large ax-space--x4"
  src="/image.png"
  style={
    Object {
      "backgroundColor": undefined,
      "height": undefined,
      "maxWidth": "100%",
      "width": undefined,
    }
  }
/>
`;

exports[`Image renders with border small 1`] = `
<img
  className="ax-image ax-image--square ax-image--border-small ax-space--x4"
  src="/image.png"
  style={
    Object {
      "backgroundColor": undefined,
      "height": undefined,
      "maxWidth": "100%",
      "width": undefined,
    }
  }
/>
`;

exports[`Image renders with defaultProps 1`] = `
<img
  className="ax-image ax-image--square ax-space--x4"
  src="/image.png"
  style={
    Object {
      "backgroundColor": undefined,
      "height": undefined,
      "maxWidth": "100%",
      "width": undefined,
    }
  }
/>
`;

exports[`Image renders with shape circle 1`] = `
<img
  className="ax-image ax-image--circle ax-space--x4"
  src="/image.png"
  style={
    Object {
      "backgroundColor": undefined,
      "height": undefined,
      "maxWidth": "100%",
      "width": undefined,
    }
  }
/>
`;

exports[`Image renders with shape rounded 1`] = `
<img
  className="ax-image ax-image--rounded ax-space--x4"
  src="/image.png"
  style={
    Object {
      "backgroundColor": undefined,
      "height": undefined,
      "maxWidth": "100%",
      "width": undefined,
    }
  }
/>
`;

exports[`Image renders with shape square 1`] = `
<img
  className="ax-image ax-image--square ax-space--x4"
  src="/image.png"
  style={
    Object {
      "backgroundColor": undefined,
      "height": undefined,
      "maxWidth": "100%",
      "width": undefined,
    }
  }
/>
`;

exports[`Image with shape circle uses height for both width and height when given 1`] = `
<img
  className="ax-image ax-image--circle ax-space--x4"
  src="/image.png"
  style={
    Object {
      "backgroundColor": undefined,
      "height": "3rem",
      "maxWidth": "100%",
      "width": "3rem",
    }
  }
/>
`;

exports[`Image with shape circle uses width for both width and height when given 1`] = `
<img
  className="ax-image ax-image--circle ax-space--x4"
  src="/image.png"
  style={
    Object {
      "backgroundColor": undefined,
      "height": "3rem",
      "maxWidth": "100%",
      "width": "3rem",
    }
  }
/>
`;
